src/webserv/FaucetWebApi.ts

```diff
@@ -213,137 +330,237 @@ export class FaucetWebApi {
           status: FaucetSessionStatus.FAILED,
           failedCode: "INTERNAL_ERROR",
           failedReason: ex.toString(),
-        }
+        };
       }
     }
 
     return sessionInfo;
   }
 
-  public async onCheckUserLimits(req: IncomingMessage, userId: string): Promise<any> {
-    if(req.method !== "GET")
+  public async onCheckUserLimits(
+    req: IncomingMessage,
+    userId: string
+  ): Promise<any> {
+    if (req.method !== "GET")
       return new FaucetHttpResponse(405, "Method Not Allowed");
 
     if (!userId) {
       return new FaucetHttpResponse(500, "userId is missing");
     }
 
+    const remoteIP = this.getRemoteAddr(req);
     const moduleManager = ServiceManager.GetService(ModuleManager);
-    const time = await moduleManager.getModule<RecurringLimitsModule>("recurring-limits")?.getTimeToNewSessionStart(userId);
-    return time ? {
-      allowed: false,
-      canBeStartedAt: time
-    } : {
-      allowed: true,
-    }
+    const recurringLimitsModule =
+      moduleManager.getModule<RecurringLimitsModule>("recurring-limits");
+    const time = await recurringLimitsModule?.getTimeToNewSessionStart(
+      userId,
+      remoteIP
+    );
+
+    return time
+      ? {
+          allowed: false,
+          canBeStartedAt: time,
+        }
+      : {
+          allowed: true,
+        };
   }
 
   public async onGetSession(sessionId: string): Promise<any> {
     let session: FaucetSession;
-    if(!sessionId || !(session = ServiceManager.GetService(SessionManager).getSession(sessionId, [FaucetSessionStatus.RUNNING]))) {
+    if (
+      !sessionId ||
+      !(session = ServiceManager.GetService(SessionManager).getSession(
+        sessionId,
+        [FaucetSessionStatus.RUNNING]
+      ))
+    ) {
       return {
         status: "unknown",
-        error: "Session not found"
+        error: "Session not found",
       };
     }
 
     let sessionInfo: IClientSessionInfo;
     try {
       sessionInfo = await session.getSessionInfo();
-    } catch(ex) {
-      if(ex instanceof FaucetError) {
+    } catch (ex) {
+      if (ex instanceof FaucetError) {
         let failedCode = ex.getCode();
-        ServiceManager.GetService(FaucetProcess).emitLog(FaucetLogLevel.INFO, `[FaucetWebApi.onGetSession]: Failed getting session info: ${failedCode}; sessionId: ${sessionId}`);
+        ServiceManager.GetService(FaucetProcess).emitLog(
+          FaucetLogLevel.INFO,
+          `[FaucetWebApi.onGetSession]: Failed getting session info: ${failedCode}; sessionId: ${sessionId}`
+        );
         return {
           status: FaucetSessionStatus.FAILED,
           failedCode,
           failedReason: ex.message,
-        }
-      }
-      else {
+        };
+      } else {
         return {
           status: FaucetSessionStatus.FAILED,
           failedCode: "INTERNAL_ERROR",
           failedReason: ex.toString(),
-        }
+        };
       }
     }
 
     return sessionInfo;
   }
 
+  private async onClaimRewardByGitcoin(
```

Comment:
What if another gitcoin claim from this user is in progress, but not finished yet?

Reply:
Good catch, will fix.